apps/system/js/view_source.js

```diff
@@ -0,0 +1,110 @@
+/* global Service */
+/* global BrowserFrame */
+/* global BaseModule */
+
+'use strict';
+
+(function() {
+  var ViewSource = function() {};
+  ViewSource.SERVICES = [
+    'viewsource'
+  ];
+
+  /**
+   * ViewSource displays source of current page in an iframe as an overlay on
+   * top of mozapps.
+   * @class ViewSource
+   * @requires Service
+   * @requires BrowserFrame
+   */
+  BaseModule.create(ViewSource, {
+    name: 'ViewSource',
+    DEBUG: false,
+    _viewer: null,
+
+    /**
+     * Entry point for module.
+     * @memberof ViewSource.prototype
+     */
+    _start: function() {
+      this._hide = this._hide.bind(this);
+    },
+
+    /**
+     * Toggle the source viewer, calling {@link ViewSource#_show} if hidden
+     * or {@link ViewSource#_hide} if shown.
+     * @memberof ViewSource.prototype
+     */
+    viewsource: function() {
+      this._viewer ? this._hide() : this._show();
+    },
+
+    /**
+     * Create and show viewer and point the iframe to the current page but using
+     * the "view-source:" scheme.
+     * @memberof ViewSource.prototype
+     */
+    _show: function() {
+      var url = 'view-source:' + Service.query('getTopMostWindow').config.url;
+      var config = {
+        url: url,
+        title: url,
+        oop: true
+      };
+      var frame = new BrowserFrame(config);
+
+      var windowWidth = Service.query('LayoutManager.width');
+      var windowHeight = Service.query('getHeightFor', frame);
+
+      var container = document.createElement('div');
+      container.classList.add('view-source');
+      container.dataset.zIndexLevel = 'view-source';
+      container.style.width = windowWidth + 'px';
+      container.style.height = 'calc(' + windowHeight + 'px - 3rem)';
+      this._viewer = container;
+
+      var header = document.createElement('gaia-header');
+      header.innerHTML = `
+        <h1 data-l10n-id="viewSource"></h1>
+        <button disabled></button>
+      `;
+      header.setAttribute('action', 'close');
+      header.addEventListener('action', this._hide);
+      container.appendChild(header);
+
+      var viewsource = frame.element;
+      container.appendChild(viewsource);
+
+      var screen = document.querySelector('#screen');
+      screen.appendChild(container);
+
+      // Delay reflow of browser frame until we have a computed height of the
+      // <gaia-header> element.
+      requestAnimationFrame(function() {
+        var headerHeight = header.clientHeight;
+        viewsource.style.height =
+          'calc(' + windowHeight + 'px - 3rem - ' + headerHeight + 'px)';
+      });
+
+      window.addEventListener('home', this._hide);
+      window.addEventListener('holdhome', this._hide);
+      window.addEventListener('sleep', this._hide);
```

Comment:
the lockscreen might also show because the screen timed out and turned off.
so you still need `lockscreen-appopened` here